### ISSUE TITLE
Swap main video dominantSpeakerChanged

### ISSUE THREAD

**Issue by carlb0329** [twilio/twilio-video.js#638]
I'm using the latest release of twilio-video.js in a Yii application.  Currently using straight javascript and jQuery to get the video chat going.  I'm able to have multiple participants join the same room.  But, when I try to listen for dominantSpeakerChanged and then swap the main video accordingly, I loose the localParticipant's video feed.  

I'm currently using the detachParticipantTracks() and attachParticipantTracks() provided in the QuickStart example.  I use those methods to detach the current speaker from he main video area, then try to reattach it to the sidebar video area, but this is not working.  

I have the listener for dominantSpeakerChanged initialized when the room is joined, which then calls my custom swapDominantSpeaker() method.

Any help would be much appreciated!

**Code to reproduce the issue:**

```js
<?php 

use frontend\assets\TwilioAsset;

TwilioAsset::register($this);

?>
<div class="row row-no-gutters" style="height:calc(100vh - 160px);">
  <div class="col-xs-12">
    <p id="room-name-display"></p>
  </div>
  <div class="col-xs-12 col-sm-9" style="">
    <div id="controls" style="height:70%;">
      <div id="preview">
        <div id="local-media"></div>
      </div>
      <div id="room-controls">
        <p id="instructions">Join an existing room:</p>
        <input id="room-name" type="text" class="form-control" style="width:200px;display:inline-block;" placeholder="Enter a room name" />
        <button id="button-join" class="btn btn-primary">Join Room</button>
        <div class="clearfix" style="margin-bottom:20px;"></div>
        <button id="button-join-new" class="btn btn-primary">Create New Room</button>
        <button id="button-leave" class="btn btn-default" style="display:none;">Leave Room</button>
        <button id="button-preview" class="btn btn-primary">Preview My Camera</button>
      </div>

    </div>
    <div id="log" style="height:30%;width:100%;overflow:scroll;max-height:150px;margin-top:20px;"></div>
  </div>
  <div class="col-xs-12 col-sm-3">
    <div id="remote-media" style="width:100%;max-width:100%;"></div>
  </div>
</div>


<script type="text/javascript">

var Video = Twilio.Video;
// var Video = require('twilio-video');

var activeRoom;
var previewTracks;
var identity;
var roomName;
var localParticipant;
var currentSpeaker; //Participant

// Attach the Tracks to the DOM.
function attachTracks(tracks, container) {
  tracks.forEach(function(track) {
    var element = track.attach()
    element.id = track.name;
    container.appendChild(element);
  });
}

// Attach the Participant's Tracks to the DOM.
function attachParticipantTracks(participant, container) {
  var tracks = getTracks(participant);
  attachTracks(tracks, container);
}

// Detach the Tracks from the DOM.
function detachTracks(tracks) {
  tracks.forEach(function(track) {
    track.detach().forEach(function(detachedElement) {
      detachedElement.remove();
    });
  });
}

// Detach the Participant's Tracks from the DOM.
function detachParticipantTracks(participant) {
  var tracks = getTracks(participant);
  detachTracks(tracks);
}

// When we are about to transition away from this page, disconnect
// from the room, if joined.
window.addEventListener('beforeunload', leaveRoomIfJoined);

// Obtain a token from the server in order to connect to the Room.
$.getJSON('/chat/get-access-token', function(json) {
  var data = json.data
  identity = data.identity;
  // document.getElementById('room-controls').style.display = 'block';
  $("#room-control").hide();

  // Bind button to join an EXISTING Room.
  document.getElementById('button-join').onclick = function() {
    roomName = $('#room-name').val();
    if (!roomName) {
      alert('Please enter a room name.');
      return;
    }

    $("#log").empty();
    log("Joining room '" + roomName + "'...");
    var connectOptions = {
      name: roomName,
      dominantSpeaker: true,
      logLevel: 'debug'
    };

    if (previewTracks) {
      connectOptions.tracks = previewTracks;
    }

    // Join the Room with the token from the server and the
    // LocalParticipant's Tracks.
    Video.connect(data.token, connectOptions).then(roomJoined, function(error) {
      log('Could not connect to Twilio: ' + error.message);
    });
  };

  //Bind button for joining a NEW Room
  document.getElementById('button-join-new').onclick = function() {
    roomName = randomId(12); //Generate a random room ID.
    $("#log").empty();
    log("Joining room '" + roomName + "'...");
    var connectOptions = {
      name: roomName,
      dominantSpeaker: true,
      logLevel: 'debug'
    };

    if (previewTracks) {
      connectOptions.tracks = previewTracks;
    }

    // Join the Room with the token from the server and the
    // LocalParticipant's Tracks.
    Video.connect(data.token, connectOptions).then(roomJoined, function(error) {
      log('Could not connect to Twilio: ' + error.message);
    });
  };
  // Bind button to leave Room.
  document.getElementById('button-leave').onclick = function() {
    log('Leaving room...');
    activeRoom.disconnect();
  };
});

// Get the Participant's Tracks.
function getTracks(participant) {
  return Array.from(participant.tracks.values()).filter(function(publication) {
    return publication.track;
  }).map(function(publication) {
    return publication.track;
  });
}

function swapDominantSpeaker(newSpeaker){
  //detach the current speaker's tracks and attach them to the remote-media div
  if(currentSpeaker){
    detachParticipantTracks(currentSpeaker);
    attachParticipantTracks(currentSpeaker, $("#remote-media"));
  }
  
  //attach the new speaker's tracks to the local-media div
  log("Dominant speaker: " + newSpeaker.identity);
  currentSpeaker = newSpeaker;
  detachParticipantTracks(newSpeaker); //First detach the tracks
  attachParticipantTracks(newSpeaker, $("#local-media")); //Then reattach them to the local-media div
}

// Successfully connected!
function roomJoined(room) {
  window.room = activeRoom = room;

  //Set the currentSpeaker to the localParticipant
  localParticipant = room.localParticipant;
  currentSpeaker = localParticipant;
  
  
  $("#room-name-display").empty().append("Room name: " + room.name);
  log("Joined as '" + identity + "'");
  $('#button-join').hide();
  $('#button-join-new').hide();
  $('#button-preview').hide();
  $('#room-name').hide();
  $('#button-leave').show();
  $("#instructions").hide();

  // Attach LocalParticipant's Tracks, if not already attached.
  var previewContainer = document.getElementById('local-media');
  if (!previewContainer.querySelector('video')) {
    attachParticipantTracks(room.localParticipant, previewContainer);
  }

  // Attach the Tracks of the Room's Participants.
  room.participants.forEach(function(participant) {
    log("Already in Room: '" + participant.identity + "'");
    var previewContainer = document.getElementById('remote-media');
    attachParticipantTracks(participant, previewContainer);
  });

  // When a Participant joins the Room, log the event.
  room.on('participantConnected', function(participant) {
    log("Joining: '" + participant.identity + "'");
  });

  // When a Participant's Track is subscribed to, attach it to the DOM.
  room.on('trackSubscribed', function(track, publication, participant) {
    log("Subscribed to " + participant.identity + "'s track: " + track.kind);
    var previewContainer = document.getElementById('remote-media');
    attachTracks([track], previewContainer);
  });

  // When a Participant's Track is unsubscribed from, detach it from the DOM.
  room.on('trackUnsubscribed', function(track, publication, participant) {
    log("Unsubscribed from " + participant.identity + "'s track: " + track.kind);
    detachTracks([track]);
  });

  // When a Participant leaves the Room, detach its Tracks.
  room.on('participantDisconnected', function(participant) {
    log("RemoteParticipant '" + participant.identity + "' left the room");
    detachParticipantTracks(participant);
  });

  //When the dominant the speaker has changed
  room.on('dominantSpeakerChanged', function(participant){
    // log("Dominant speaker: " + participant);
    if(participant){
      swapDominantSpeaker(participant);
    }else{
      swapDominantSpeaker(room.localParticipant);
    }
  });

  // Once the LocalParticipant leaves the room, detach the Tracks
  // of all Participants, including that of the LocalParticipant.
  room.on('disconnected', function() {
    log('Left');
    if (previewTracks) {
      previewTracks.forEach(function(track) {
        track.stop();
      });
      previewTracks = null;
    }
    detachParticipantTracks(room.localParticipant);
    room.participants.forEach(detachParticipantTracks);
    activeRoom = null;
    $('#button-join').show();
    $('#button-join-new').show();
    $('#room-name').show();
    $('#button-preview').show();
    $("#instructions").show();
    $('#button-leave').hide();
    $("#room-name-display").empty();
  });
}

// Preview LocalParticipant's Tracks.
document.getElementById('button-preview').onclick = function() {
  var localTracksPromise = previewTracks
    ? Promise.resolve(previewTracks)
    : Video.createLocalTracks();

  localTracksPromise.then(function(tracks) {
    window.previewTracks = previewTracks = tracks;
    var previewContainer = document.getElementById('local-media');
    if (!previewContainer.querySelector('video')) {
      attachTracks(tracks, previewContainer);
      $('#button-preview').hide();
    }
  }, function(error) {
    console.error('Unable to access local media', error);
    log('Unable to access Camera and Microphone');
  });
};

// Activity log.
function log(message) {
  var logDiv = document.getElementById('log');
  logDiv.innerHTML += '<p>&gt;&nbsp;' + message + '</p>';
  logDiv.scrollTop = logDiv.scrollHeight;
}

// Leave Room.
function leaveRoomIfJoined() {
  if (activeRoom) {
    activeRoom.disconnect();
  }
}
</script>
```

**Expected behavior:**

Videos should swap accordingly to which video has the dominant speaker.

**Actual behavior:**

The local participant video disappears and does not re-attach to the window correctly.

**Software versions:**

 - [x] Browser(s):
Safari and Chrome
 - [x] Operating System:
Mac OS X and iOS
 - [x] twilio-video.js:
2.0.0-rc23
 - [x] Third-party libraries (e.g., Angular, React, etc.):
Yii PHP framework

**Comment by manjeshbhargav** [comment]
Hi @carlb0329 ,

Thanks for writing in with your question. It's been a while since I've used jQuery, so please forgive me if I'm asking the wrong question. The `container` argument in `attachTracks()` should be a raw HTMLElement. Maybe you should pass `$('#local-media').get(0)` to it?

On an unrelated note, we don't recommend using internal release candidates in your production applications. You can use the latest release 2.0.0-beta9 which also supports the Dominant Speaker API.

Please let me know if this helps.

Thanks,

Manjesh Malavalli
JSDK Team

**Comment by carlb0329** [comment]
@manjeshbhargav ,

Ah! I think that worked.  My videos are no longer disappearing and seem to be attaching to the correct DOM elements now. Good catch!

I will make sure to update to the release version.  On a related note, I'm using Composer to install the npm-asset/twilio-video library and have had some issues configuring the composer.json to get the correct version of twilio-video.  I had  ``` "npm-asset/twilio-video":"^2.0" ``` in my composer.json and it had installed version ``` "2.0.0-rc23" ``` of the twilio-video library, but you mentioned I should be using the 2.0.0-beta9 release.  Any idea on what the composer.json entry should be?  No worries if not.

**Comment by manjeshbhargav** [comment]
Hi @carlb0329 ,

I'm not familiar with composer, but I think for now you can specify the exact version `2.0.0-beta9`.
I'll close this issue for now. Please feel free to open a new one if you experience any other issues.

Thanks,

Manjesh Malavalli
JSDK Team
